### ISSUE TITLE
Problem with saving image from clipboard

### ISSUE THREAD

**Issue by sallyruthstruik** [python-pillow/Pillow#1293]
Hello,

I have Windows 8 x64 system and installed Pillow version 2.8.2
I'm trying to grab image from clipboard. I press "Alt+PrtSc" and then call grabclipboard function, and get such error:
```{python}
ImageGrab.grabclipboard()
---------------------------------------------------------------------------
IOError                                   Traceback (most recent call last)
<ipython-input-2-c8274e888e6c> in <module>()
      1 from PIL import ImageGrab
      2 
----> 3 ImageGrab.grabclipboard()

C:\Anaconda\lib\site-packages\PIL\ImageGrab.pyc in grabclipboard()
     49         from PIL import BmpImagePlugin
     50         import io
---> 51         return BmpImagePlugin.DibImageFile(io.BytesIO(data))
     52     return data

C:\Anaconda\lib\site-packages\PIL\ImageFile.pyc in __init__(self, fp, filename)
     95 
     96         try:
---> 97             self._open()
     98         except IndexError as v:  # end of data
     99             if Image.DEBUG > 1:

C:\Anaconda\lib\site-packages\PIL\BmpImagePlugin.pyc in _open(self)
    204 
    205     def _open(self):
--> 206         self._bitmap()
    207 
    208 #

C:\Anaconda\lib\site-packages\PIL\BmpImagePlugin.pyc in _bitmap(self, header, offset)
    145                     raw_mode = MASK_MODES[(file_info['bits'], file_info['rgb_mask'])]
    146                 else:
--> 147                     raise IOError("Unsupported BMP bitfields layout")
    148             else:
    149                 raise IOError("Unsupported BMP bitfields layout")

IOError: Unsupported BMP bitfields layout
```

But if I call ImageGrab.grab function, screenshot(of whole screen) is saved properly.

**Comment by sallyruthstruik** [comment]
UPD It's version problem, when I installed Pillow version 2.6.0 everything worked ok

**Comment by hugovk** [comment]
Reproducible with Pilllow 2.8.0, 2.8.1, 2.8.2. 
Not reproducible with Pillow 2.6.0, 2.7.0.

```
E:\stufftodelete>cat test.py
from PIL import ImageGrab
ImageGrab.grabclipboard()

E:\stufftodelete>pip freeze | grep illow
Pillow==2.8.2

E:\stufftodelete>python test.py
Traceback (most recent call last):
  File "test.py", line 2, in <module>
    ImageGrab.grabclipboard()
  File "C:\Python27\lib\site-packages\PIL\ImageGrab.py", line 51, in grabclipboard
    return BmpImagePlugin.DibImageFile(io.BytesIO(data))
  File "C:\Python27\lib\site-packages\PIL\ImageFile.py", line 97, in __init__
    self._open()
  File "C:\Python27\lib\site-packages\PIL\BmpImagePlugin.py", line 206, in _open
    self._bitmap()
  File "C:\Python27\lib\site-packages\PIL\BmpImagePlugin.py", line 147, in _bitmap
    raise IOError("Unsupported BMP bitfields layout")
IOError: Unsupported BMP bitfields layout

E:\stufftodelete>pip install pillow==2.8.1
Collecting pillow==2.8.1
  Using cached Pillow-2.8.1-cp27-none-win_amd64.whl
Installing collected packages: pillow
  Found existing installation: Pillow 2.8.2
    Uninstalling Pillow-2.8.2:
      Successfully uninstalled Pillow-2.8.2
Successfully installed pillow-2.8.1

E:\stufftodelete>python test.py
Traceback (most recent call last):
  File "test.py", line 2, in <module>
    ImageGrab.grabclipboard()
  File "C:\Python27\lib\site-packages\PIL\ImageGrab.py", line 51, in grabclipboard
    return BmpImagePlugin.DibImageFile(io.BytesIO(data))
  File "C:\Python27\lib\site-packages\PIL\ImageFile.py", line 97, in __init__
    self._open()
  File "C:\Python27\lib\site-packages\PIL\BmpImagePlugin.py", line 206, in _open
    self._bitmap()
  File "C:\Python27\lib\site-packages\PIL\BmpImagePlugin.py", line 147, in _bitmap
    raise IOError("Unsupported BMP bitfields layout")
IOError: Unsupported BMP bitfields layout

E:\stufftodelete>pip install pillow==2.8.0
Collecting pillow==2.8.0
  Downloading Pillow-2.8.0-cp27-none-win_amd64.whl (1.3MB)
    100% |################################| 1.3MB 386kB/s
Installing collected packages: pillow
  Found existing installation: Pillow 2.8.1
    Uninstalling Pillow-2.8.1:
      Successfully uninstalled Pillow-2.8.1
Successfully installed pillow-2.8.0

E:\stufftodelete>python test.py
Traceback (most recent call last):
  File "test.py", line 2, in <module>
    ImageGrab.grabclipboard()
  File "C:\Python27\lib\site-packages\PIL\ImageGrab.py", line 51, in grabclipboard
    return BmpImagePlugin.DibImageFile(io.BytesIO(data))
  File "C:\Python27\lib\site-packages\PIL\ImageFile.py", line 97, in __init__
    self._open()
  File "C:\Python27\lib\site-packages\PIL\BmpImagePlugin.py", line 206, in _open
    self._bitmap()
  File "C:\Python27\lib\site-packages\PIL\BmpImagePlugin.py", line 147, in _bitmap
    raise IOError("Unsupported BMP bitfields layout")
IOError: Unsupported BMP bitfields layout

E:\stufftodelete>pip install pillow==2.7.0
Collecting pillow==2.7.0
  Downloading Pillow-2.7.0-cp27-none-win_amd64.whl (1.3MB)
    100% |################################| 1.3MB 393kB/s
Installing collected packages: pillow
  Found existing installation: Pillow 2.8.0
    Uninstalling Pillow-2.8.0:
      Successfully uninstalled Pillow-2.8.0
Successfully installed pillow-2.7.0

E:\stufftodelete>python test.py

E:\stufftodelete>pip install pillow==2.6.0
Collecting pillow==2.6.0
  Downloading Pillow-2.6.0-cp27-none-win_amd64.whl (1.3MB)
    100% |################################| 1.3MB 405kB/s
Installing collected packages: pillow
  Found existing installation: Pillow 2.7.0
    Uninstalling Pillow-2.7.0:
      Successfully uninstalled Pillow-2.7.0
Successfully installed pillow-2.6.0

E:\stufftodelete>python test.py
```

**Comment by hugovk** [comment]
Pillow 2.7.0 was released on 1st January this year, and 2.8.0 was released on 1st April.

PR https://github.com/python-pillow/Pillow/pull/1125 may have caused it. Ping @artscoop.

It'd be great to get this fixed for 2.9.0, due out on 1st July.

The good news is we now have a Windows-based CI to run unit tests, so if it's possible to simulate Alt+PrtSc (or just PrtSc, which also triggers it) on AppVeyor then this should definitely be unit-tested.

**Comment by radarhere** [comment]
Running OS X, I can reproduce the error using one of the files in the test suite - [Tests/images/bmp/g/rgb32bf.bmp](https://github.com/python-pillow/Pillow/blob/master/Tests/images/bmp/g/rgb32.bmp)

Interestingly though, it's [listed as unsupported](https://github.com/python-pillow/Pillow/blob/master/Tests/test_bmp_reference.py#L82).

**Comment by wiredfool** [comment]
Unfortunately, the ImageGrab functions don't work on the windows CI, likely because there's something screwy with the graphics context on a headless vm.

**Comment by hugovk** [comment]
Notes:

This can be partially fixed by adding this `elif` near [here](https://github.com/python-pillow/Pillow/blob/master/PIL/BmpImagePlugin.py#L146):
```python
                elif file_info['bits'] == 32 and file_info['rgb_mask'] == (0xff0000, 0xff00, 0xff):
                    pass
                else:
                    raise IOError("Unsupported BMP bitfields layout")
            else:
                raise IOError("Unsupported BMP bitfields layout")
```

That will return an image, but the resulting image is truncated. It can be shown or saved only if `LOAD_TRUNCATED_IMAGES` is true, but the image is one pixel shifted left and missing the top line.

This is because, for example, `self.tile` is:
`[('raw', (0, 0, 3360, 1050), 56L, ('BGRX', 13440, -1))]` 
instead of:
`[('raw', (0, 0, 3360, 1050), 52L, ('BGRX', 13440, -1))]`

That third value in both cases is `offset or self.fp.tell()`, the problem with 2.8.0 is that an [extra read](https://github.com/python-pillow/Pillow/blob/master/PIL/BmpImagePlugin.py#L112) had been made to load in a `file_info['a_mask']`, which doesn't make sense for this format, and putting the file pointer at the wrong place.

We should only be reading that alpha value when needed.

**Comment by techtonik** [comment]
The regression appears with 3.0.0 too.

**Comment by techtonik** [comment]
So, any news on that?

**Comment by artscoop** [comment]
Ok I understand what is going on.
I rewrote part of the BMP loading process.
And I made sure it worked on supported bitmap files, fixed bogus behaviour from previous versions, but it did not have to work with unsupported files.
However @radarhere , the cause of the error you've been experiencing with this file may be... let's say, different.

**Comment by artscoop** [comment]
Now I remember: the file rgb32bf.bmp is unsupported because it has a bitmask not supported by Pillow. Supported masks for 32-bit true color images have classic boundaries, with RGB channels in RGB or BGR order.
Here we have an esoteric RGB/A mask (the image is read correctly by XNView though), which has no matching dialect in PIL definitions for unpacking pixels.
Therefore I have been able to load this image, but with wrong colors (for instance, we have a problem with B and G channels)

**Comment by techtonik** [comment]
This bug can be closed. This works with 3.2.0:
```
>>> from PIL import ImageGrab
>>> ImageGrab.grabclipboard()
>>>
```

**Comment by techtonik** [comment]
It can be renamed "ImageGrab.grabclipboard() problem on Windows", because there are other issues there.

**Comment by hugovk** [comment]
This bug is still present: I can reproduce it as described in the original report with Pillow 3.2.0 (Python 2.7, Windows 7).

**Comment by techtonik** [comment]
@hugovk can you confirm it with the following actions?

1 open mspaint
2 alt+prtscr
3 run the code
```python
from PIL import ImageGrab
im = ImageGrab.grabclipboard()
if im is None:
  print('Error: No image data in clipboard')
```
Python 2.7, 32-bit Vista, Pillow 3.2.0 works.

**Comment by hugovk** [comment]
@techtonik 

* If I alt+prtscr the cmd.exe window, I get `IOError: Unsupported BMP bitfields layout`.

* If I alt+prtscr the Paint window, I get an `im` (e.g. `<PIL.BmpImagePlugin.DibImageFile image mode=RGBA size=886x754 at 0x2A7FE10>`) but `im.show()` gives `IOError: image file is truncated (3540 bytes not processed)`

**Comment by techtonik** [comment]
@hugovk yep, the same here, so looks like this issue is only about console window.

**Comment by wiredfool** [comment]
What is the bitfield layout? If it's sane, there's no reason that we can't support it.

**Comment by techtonik** [comment]
@wiredfool any ideas where to look for it? Let me update the Traceback for 3.2.0 first.
```
Traceback (most recent call last):
  File "code.py", line 2, in <module>
    im = ImageGrab.grabclipboard()
  File "C:\Python27\lib\site-packages\PIL\ImageGrab.py", line 60, in grabclipboard
    return BmpImagePlugin.DibImageFile(io.BytesIO(data))
  File "C:\Python27\lib\site-packages\PIL\ImageFile.py", line 97, in __init__
    self._open()
  File "C:\Python27\lib\site-packages\PIL\BmpImagePlugin.py", line 211, in _open
    self._bitmap()
  File "C:\Python27\lib\site-packages\PIL\BmpImagePlugin.py", line 152, in _bitmap
    raise IOError("Unsupported BMP bitfields layout")
IOError: Unsupported BMP bitfields layout
```

**Comment by techtonik** [comment]
Printing `file_info` contents just before `raise`:
```python
{
    'direction': -1,
    'r_mask': 16711680,
    'compression': 3,
    'pixels_per_meter': (0, 0),
    'palette_padding': 4,
    'g_mask': 65280,
    'a_mask': 6908265,
    'rgba_mask': (16711680, 65280, 255, 6908265),
    'header_size': 40,
    'height': 736,
    'width': 1066,
    'colors': 4294967296L,
    'b_mask': 255,
    'rgb_mask': (16711680, 65280, 255),
    'planes': 1,
    'y_flip': False,
    'bits': 32,
    'data_size': 3138304,
}
```

**Comment by wiredfool** [comment]
ok, that's a WTF. 

The alpha mask is 0x696969, so it's a BRG[WTF] image??

**Comment by techtonik** [comment]
Screenshot of a console window:

![image](https://cloud.githubusercontent.com/assets/515889/14314532/714f11ea-fc00-11e5-9c94-4c4b35f0968a.png)

**Comment by wiredfool** [comment]
Try this:
```
diff --git a/PIL/BmpImagePlugin.py b/PIL/BmpImagePlugin.py
index e398445..f7cc203 100644
--- a/PIL/BmpImagePlugin.py
+++ b/PIL/BmpImagePlugin.py
@@ -142,6 +142,10 @@ class BmpImageFile(ImageFile.ImageFile):
                 (16, (0xf800, 0x7e0, 0x1f)): "BGR;16",
                 (16, (0x7c00, 0x3e0, 0x1f)): "BGR;15"
             }
+            # work around really strange alpha reported in windows screenshot bmps.
+            if file_info['bits'] == 32 and file_info['a_mask'] == 0x696969:
+                file_info['bits'] = 24
+                
             if file_info['bits'] in SUPPORTED:
                 if file_info['bits'] == 32 and file_info['rgba_mask'] in SUPPORTED[file_info['bits']]:
                     raw_mode = MASK_MODES[(file_info['bits'], file_info['rgba_mask'])]
```

**Comment by techtonik** [comment]
The mask of the image is:
```
(0xff0000, 0xff00, 0xff, 0x696969)
```
while supported ones for 32 bits are:
```
(0xff0000, 0xff00, 0xff, 0x0)
(0xff0000, 0xff00, 0xff, 0xff000000)
(0x0, 0x0, 0x0, 0x0)
```
I can't find any references of that magic number. Maybe Pillow gets that number wrong?

**Comment by wiredfool** [comment]
That magic number makes no sense at all from a masking point of view.

**Comment by techtonik** [comment]
@wiredfool this is the result

![cmd](https://cloud.githubusercontent.com/assets/515889/14314777/018f4706-fc02-11e5-9001-a08461ebc2ee.png)

**Comment by wiredfool** [comment]
That looks like aliasing.  so, probably BGRX instead of BGR

Try this:
```
diff --git a/PIL/BmpImagePlugin.py b/PIL/BmpImagePlugin.py
index e398445..e3b3399 100644
--- a/PIL/BmpImagePlugin.py
+++ b/PIL/BmpImagePlugin.py
@@ -130,13 +130,15 @@ class BmpImageFile(ImageFile.ImageFile):
         # ----------------- Process BMP with Bitfields compression (not palette)
         if file_info['compression'] == self.BITFIELDS:
             SUPPORTED = {
-                32: [(0xff0000, 0xff00, 0xff, 0x0), (0xff0000, 0xff00, 0xff, 0xff000000), (0x0, 0x0, 0x0, 0x0)],
+                32: [(0xff0000, 0xff00, 0xff, 0x0), (0xff0000, 0xff00, 0xff, 0xff000000),
+                     (0xff0000, 0xff00, 0xff, 0x696969), (0x0, 0x0, 0x0, 0x0)],
                 24: [(0xff0000, 0xff00, 0xff)],
                 16: [(0xf800, 0x7e0, 0x1f), (0x7c00, 0x3e0, 0x1f)]
             }
             MASK_MODES = {
                 (32, (0xff0000, 0xff00, 0xff, 0x0)): "BGRX",
                 (32, (0xff0000, 0xff00, 0xff, 0xff000000)): "BGRA",
+                (32, (0xff0000, 0xff00, 0xff, 0x696969)): "BGRX",
                 (32, (0x0, 0x0, 0x0, 0x0)): "BGRA",
                 (24, (0xff0000, 0xff00, 0xff)): "BGR",
                 (16, (0xf800, 0x7e0, 0x1f)): "BGR;16",
```

**Comment by techtonik** [comment]
@wiredfool yay, this works (with `ImageFile.LOAD_TRUNCATED_IMAGES = True`, which is another issue - #1810).

**Comment by wiredfool** [comment]
Great!.  Could you dump the data in ImageGrab.grabclipboard() to a file for use in a test?

**Comment by techtonik** [comment]
@wiredfool how to dump?

**Comment by wiredfool** [comment]
just before the return BmpImagePlugin line:

with open('test_file.bmp', 'wb') as f:
    f.write(data)

**Comment by techtonik** [comment]
I used `im._dump('console.dib')` after `grabclipboard` to get [console.dib.zip](https://github.com/python-pillow/Pillow/files/206348/console.dib.zip). Let me try with `f.write()`

**Comment by wiredfool** [comment]
Image._dump saves a PPM file, I want the raw bytes so that we can get a good test out of it.

**Comment by techtonik** [comment]
Here you go - [console.raw.zip](https://github.com/python-pillow/Pillow/files/206355/console.raw.zip)
.

**Comment by wiredfool** [comment]
Thanks. I think I've found the bug, In the DIB file, it looks like they're only writing in the three masks, and we're reading 4. The 0x696969 is actually the color of the first pixel. Which is why the image is truncated, since the first pixel is being read as an alpha mask.

**Comment by techtonik** [comment]
That's something mentioned by @hugovk in https://github.com/python-pillow/Pillow/issues/1293#issuecomment-115237352 that I could not understand.

**Comment by wiredfool** [comment]
From https://msdn.microsoft.com/en-us/library/windows/desktop/dd183376(v=vs.85).aspx , in the section on the 40 byte headers, which is the type that we're using here:

> BI_BITFIELDS	Specifies that the bitmap is not compressed and that the color table consists of three DWORD color masks that specify the red, green, and blue components, respectively, of each pixel. This is valid when used with 16- and 32-bpp bitmaps.

So, when we have a 40 byte header, we can't read the 4 components of the mask.

**Comment by wiredfool** [comment]
Ok, I think this patch should do it, where the previous one would only work if the first pixel color was 0x696969.

```
diff --git a/PIL/BmpImagePlugin.py b/PIL/BmpImagePlugin.py
index e398445..a3e39c1 100644
--- a/PIL/BmpImagePlugin.py
+++ b/PIL/BmpImagePlugin.py
@@ -109,7 +109,11 @@ class BmpImageFile(ImageFile.ImageFile):
                         for idx, mask in enumerate(['r_mask', 'g_mask', 'b_mask', 'a_mask']):
                             file_info[mask] = i32(header_data[36+idx*4:40+idx*4])
                     else:
-                        for mask in ['r_mask', 'g_mask', 'b_mask', 'a_mask']:
+                        # 40 byte headers only have the three components in the bitfields masks,
+                        # ref: https://msdn.microsoft.com/en-us/library/windows/desktop/dd183376(v=vs.85).aspx
+                        # See also https://github.com/python-pillow/Pillow/issues/1293
+                        file_info['a_mask'] = 0x0
+                        for mask in ['r_mask', 'g_mask', 'b_mask']:
                             file_info[mask] = i32(read(4))
                     file_info['rgb_mask'] = (file_info['r_mask'], file_info['g_mask'], file_info['b_mask'])
                     file_info['rgba_mask'] = (file_info['r_mask'], file_info['g_mask'], file_info['b_mask'], file_info['a_mask'])
```

**Comment by wiredfool** [comment]
Hmm. Looks like there is an alpha channel involved, and it's inverted -- e.g. 00 is solid and FF is transparent.

**Comment by techtonik** [comment]
This patch works too.

**Comment by techtonik** [comment]
I can confirm this is fixed for me.

**Comment by jenstroeger** [comment]
I haven't built this from source, so when will the DIB/DMP fix make it into the Python module and become available through pip?

**Comment by hugovk** [comment]
Releases are quarterly; next one is due for 1st July.

**Comment by wiredfool** [comment]
Note that the current fix isn't actually correct, I believe that it's misinterpreting some active mask as transparency.

**Comment by jenstroeger** [comment]
I noticed black backgrounds in bmp to png converted images... ;-)

**Comment by wiredfool** [comment]
I need to dig a bit more about the actual definition of what a DIB file is, and what the data in the spot that would normally be an alpha channel actually corresponds to.  I think that the fix is actually correct for the `RGB` channels, for what that's worth.

**Comment by wiredfool** [comment]
BMP does not depend on any external library.  (This particular version is essentially a header + raw image bytes)

**Comment by jenstroeger** [comment]
Glad to hear there are no external dependencies for BMP. I noticed the DIB problems because it seems to be the internal bitmap representation for graphic shapes in OpenOffice (see [XBitmap](https://www.openoffice.org/api/docs/common/ref/com/sun/star/awt/XBitmap.html) here).

Microsoft has an overview of DIBs [here](https://msdn.microsoft.com/en-us/library/windows/desktop/dd183562(v=vs.85).aspx) and a more historical piece [here](https://msdn.microsoft.com/en-us/library/ms969901.aspx), and the actual “Bitmap Reference” is [here](https://msdn.microsoft.com/en-us/library/dd183388(v=vs.85).aspx). I hope this helps.

**Comment by wiredfool** [comment]
I've seen most of those links. What none of them are saying is the interpretation of the extra field of an RGBQuad, other than 'reserved'.  There's possibly something newer that's done something documented with it, or we need to simply ignore it and just use the 'RGB' section of the image.

**Comment by techtonik** [comment]
@wiredfool lets close this ticket and move the discussion to #1811. Because saved .bmp is fine. There might be a bug in conversion routines.

**Comment by wiredfool** [comment]
https://github.com/python-pillow/Pillow/issues/1293#issuecomment-210117566

**Comment by techtonik** [comment]
@wiredfool "IOError: Unsupported BMP bitfields layout" problem is fixed, and it becomes hard to follow this ticket.

**Comment by wiredfool** [comment]
Respinning this with #1847.
